numpy/core/src/umath/ufunc_object.c

```diff
@@ -2216,25 +2230,50 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
     /* Get all the arguments */
     retval = get_ufunc_arguments(ufunc, args, kwds,
                 op, &order, &casting, &extobj,
-                &type_tup, &subok, NULL, &axes);
+                &type_tup, &subok, NULL, &axes, &keepdims);
     if (retval < 0) {
         goto fail;
     }
-
+    /*
+     * If keepdims was passed in (and thus changed from the initial value
+     * on top), check the gufunc is suitable.
+     */
+    if (keepdims != -1) {
+        for (i = 0; i < nop; i++) {
+            if (ufunc->core_num_dims[i] != (i < nin? 1 : 0)) {
+                PyErr_Format(PyExc_TypeError,
+                    "%s: keepdims can only be used when all inputs have "
+                    "single core dimensions and all outputs have no core "
+                    "dimensions, but signature %s requires that %s %d has "
+                    "%d core dimensions",
+                    ufunc_name, ufunc->core_signature,
+                    i < nin ? "input" : "output", i < nin ? i : i - nin,
+                     ufunc->core_num_dims[i]);
+                retval = -1;
+                goto fail;
+            }
+        }
+    }
+    else {
+        /* keepdims was not set; treat as false below */
+        keepdims = 0;
+    }
     /*
      * Check that operands have the minimum dimensions required.
      * (Just checks core; broadcast dimensions are tested by the iterator.)
      */
     for (i = 0; i < nop; i++) {
-        if (op[i] != NULL && PyArray_NDIM(op[i]) < ufunc->core_num_dims[i]) {
+        int num_dims = keepdims && i >= nin? 1: ufunc->core_num_dims[i];
+
+        if (op[i] != NULL && PyArray_NDIM(op[i]) < num_dims) {
             PyErr_Format(PyExc_ValueError,
                          "%s: %s operand %d does not have enough "
                          "dimensions (has %d, gufunc core with "
                          "signature %s requires %d)",
-                         ufunc_get_name_cstr(ufunc),
-                         i < nin ? "Input" : "Output",
+                         ufunc_name, i < nin ? "Input" : "Output",
```

Comment:
This was better with the two ternaries line wrapped to match up

Reply:
OK, now have all arguments on their own line.